src/libponyc/ast/ast.c

```diff
@@ -1435,6 +1587,33 @@ void ast_free_unattached(ast_t* ast)
     ast_free(ast);
 }
 
+bool ast_is_frozen(ast_t* ast)
+{
+  pony_assert(ast != NULL);
+
+#if !defined(NDEBUG) || !defined(PONY_NO_ASSERT)
+  return ast->frozen;
+#else
+  return false;
+#endif
+}
+
+void ast_freeze(ast_t* ast)
+{
+#if !defined(NDEBUG) || !defined(PONY_NO_ASSERT)
+  if((ast == NULL) || ast->frozen)
+    return;
+
+  ast->frozen = true;
+  token_freeze(ast->t);
+
+  for(ast_t* child = ast->child; child != NULL; child = ast_sibling(child))
+    ast_freeze(child);
+
+  ast_freeze(ast->annotation_type);
```

Comment:
This won't freeze the type if there is annotation *and* a type attached.

Reply:
Wouldn't the annotation call `ast_freeze` on its `annotation_type` field, which will freeze the type?

Reply:
Ah, right. My mistake.